app/controllers/vaults_controller.rb

```diff
@@ -89,13 +90,21 @@ def destroy
   end
 
   private
+
     # Use callbacks to share common setup or constraints between actions.
     def set_vault
-      @vault = Vault.find(params[:id])
+      @vault = Vault.find_by_id(params[:id])
+      redirect_to root_url if @vault.nil?
+    end
+
+    # Check is the given username is owned by the vault's user
+    def authenticate_username
+      vault = Vault.find(params[:id])
+      redirect_to root_url unless vault.user_username == params[:username]
```

Comment:
Could `authenticate_username` and `set_vault` not be merged?

If not, using the instance variable `@vault` set by `set_vault` within `authenticate_username` makes more sense than getting the same data from the DB again.